internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,214 @@
+	PullSecretUpdatesChannel chan event.GenericEvent
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs/status,verbs=get;update;patch
+
+func (r *InstallEnvReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.ensureISO(ctx, installEnv)
+}
+
+// ensureISO generates ISO for the cluster if needed and will update the condition Reason and Message accordingly.
+// It returns a result that includes ISODownloadURL.
+func (r *InstallEnvReconciler) ensureISO(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var imageInfo *models.ImageInfo
+	var updatedCluster *common.Cluster
+	imageState := ImageState{
+		State:     ImageCreationError,
+		StateInfo: adiiov1alpha1.ImageStateFailedToCreate,
+	}
+
+	kubeKey := types.NamespacedName{
+		Name:      installEnv.Spec.ClusterRef.Name,
+		Namespace: installEnv.Spec.ClusterRef.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		imageState.State = ImageBeingCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.State = ImageCreated
```

Comment:
Not sure why `ImageState` is needed?

Reply:
`ImageState` just aggregates `State` and `StateInfo` to pass them both around